File: backend/src/utils/validation.py
"""
Validation utilities
Provides validators for email, password, and todo fields
"""
import re
from typing import Tuple


def validate_email(email: str) -> Tuple[bool, str]:
    """
    Validate email format

    Args:
        email: Email address to validate

    Returns:
        (is_valid, error_message) tuple
    """
    if not email or len(email.strip()) == 0:
        return False, "Email is required"

    if len(email) > 255:
        return False, "Email must be at most 255 characters"

    # Basic email regex pattern
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'

    if not re.match(pattern, email):
        return False, "Invalid email format"

    return True, ""


def validate_password(password: str) -> Tuple[bool, str]:
    """
    Validate password strength

    Requirements:
    - At least 8 characters
    - At most 128 characters

    Args:
        password: Password to validate

    Returns:
        (is_valid, error_message) tuple
    """
    if not password:
        return False, "Password is required"

    if len(password) < 8:
        return False, "Password must be at least 8 characters"

    if len(password) > 128:
        return False, "Password must be at most 128 characters"

    return True, ""


def validate_todo_title(title: str) -> Tuple[bool, str]:
    """
    Validate todo title

    Requirements:
    - Required (non-empty)
    - At most 200 characters

    Args:
        title: Todo title to validate

    Returns:
        (is_valid, error_message) tuple
    """
    if not title or len(title.strip()) == 0:
        return False, "Title is required"

    if len(title) > 200:
        return False, "Title must be at most 200 characters"

    return True, ""


def validate_todo_description(description: str) -> Tuple[bool, str]:
    """
    Validate todo description

    Requirements:
    - Optional (can be empty)
    - At most 2000 characters if provided

    Args:
        description: Todo description to validate

    Returns:
        (is_valid, error_message) tuple
    """
    if description and len(description) > 2000:
        return False, "Description must be at most 2000 characters"

    return True, ""

File: frontend/src/lib/utils/validation.ts
/**
 * Frontend validation utilities
 * Provides client-side validation for user input
 */

export interface ValidationResult {
  isValid: boolean;
  error?: string;
}

/**
 * Validate email format
 */
export function validateEmail(email: string): ValidationResult {
  if (!email || email.trim().length === 0) {
    return { isValid: false, error: "Email is required" };
  }

  if (email.length > 255) {
    return { isValid: false, error: "Email must be at most 255 characters" };
  }

  const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;

  if (!emailRegex.test(email)) {
    return { isValid: false, error: "Invalid email format" };
  }

  return { isValid: true };
}

/**
 * Validate password strength
 */
export function validatePassword(password: string): ValidationResult {
  if (!password) {
    return { isValid: false, error: "Password is required" };
  }

  if (password.length < 8) {
    return { isValid: false, error: "Password must be at least 8 characters" };
  }

  if (password.length > 128) {
    return { isValid: false, error: "Password must be at most 128 characters" };
  }

  return { isValid: true };
}

/**
 * Validate todo title
 */
export function validateTodoTitle(title: string): ValidationResult {
  if (!title || title.trim().length === 0) {
    return { isValid: false, error: "Title is required" };
  }

  if (title.length > 200) {
    return { isValid: false, error: "Title must be at most 200 characters" };
  }

  return { isValid: true };
}

/**
 * Validate todo description
 */
export function validateTodoDescription(description: string): ValidationResult {
  if (description && description.length > 2000) {
    return { isValid: false, error: "Description must be at most 2000 characters" };
  }

  return { isValid: true };
}
